hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/utils/SlidingWindow.java

```diff
@@ -0,0 +1,162 @@
+ *     automatically removed
+ * <li>Size-based: The window maintains at most windowSize latest events, removing
+ *     older events when this limit is exceeded
+ * </ul>
+ *
+ * The window is considered full when the number of non-expired events exceeds
+ * the specified window size. Events are automatically pruned based on both
+ * their age and the maximum size constraint.
+ */
+public class SlidingWindow {
+  private final Object lock = new Object();
+  private final int windowSize;
+  private final Deque<Long> timestamps;
+  private final long expiryDurationMillis;
+  private final Clock clock;
+
+  /**
+   * Default constructor that uses a monotonic clock.
+   *
+   * @param windowSize     the maximum number of events that are tracked
+   * @param expiryDuration the duration after which an entry in the window expires
+   */
+  public SlidingWindow(int windowSize, Duration expiryDuration) {
+    this(windowSize, expiryDuration, new MonotonicClock());
+  }
+
+  /**
+   * Constructor with a custom clock for testing.
+   *
+   * @param windowSize     the maximum number of events that are tracked
+   * @param expiryDuration the duration after which an entry in the window expires
+   * @param clock          the clock to use for time measurements
+   */
+  public SlidingWindow(int windowSize, Duration expiryDuration, Clock clock) {
+    if (windowSize < 0) {
+      throw new IllegalArgumentException("Window size must be greater than 0");
+    }
+    if (expiryDuration.isNegative() || expiryDuration.isZero()) {
+      throw new IllegalArgumentException("Expiry duration must be greater than 0");
+    }
+    this.windowSize = windowSize;
+    this.expiryDurationMillis = expiryDuration.toMillis();
+    this.clock = clock;
+    // We limit the initial queue size to 100 to control the memory usage
+    this.timestamps = new ArrayDeque<>(Math.min(windowSize + 1, 100));
+  }
+
+  public void add() {
+    synchronized (lock) {
+      if (isExceeded()) {
+        timestamps.remove();
+      }
+
+      timestamps.add(getCurrentTime());
+    }
+  }
+
+  /**
+   * Checks if the sliding window has exceeded its maximum size.
+   * This is useful to track if we have encountered more events than the window's defined limit.
+   * @return true if the number of tracked timestamps in the sliding window
+   *         exceeds the specified window size, false otherwise.
+   */
+  public boolean isExceeded() {
+    synchronized (lock) {
+      removeExpired();
+      return timestamps.size() > windowSize;
+    }
+  }
+
+  /**
+   * Returns the current number of events that are tracked within the sliding window.
+   * This method ensures that expired events are removed before computing the count.
+   *
+   * @return the number of valid timestamps currently in the sliding window
+   */
+  public int getNumEvents() {
+    synchronized (lock) {
+      removeExpired();
+      return timestamps.size();
+    }
+  }
```

Comment:
I think it would be more intuitive for the number of events returned to never be larger than the window size. The window is only reliably tracking event counts that fall within the size anyways, since we only keep one event beyond to check if it was exceeded.
```suggestion
  public int getNumEventsInWindow() {
    synchronized (lock) {
      removeExpired();
      return Math.min(timestamps.size(), windowSize);
    }
  }
```

Reply:
The `getNumEvents` is used only for testing. I have added an annotation for clarifying this. Regardless of the window size, We do want to know to know the exact number of elements in the queue to perform validations in the test.

Reply:
I had already made that change for the [upcoming PR](https://github.com/ptlrs/ozone/blob/b1df449029788d0f1c8ee492a6bcf21ede6a0f90/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/StorageVolume.java#L705-L709).
I am using `getWindowSize()` for this. If we want to specify the exact items I have pushed the new method as well.